examples/alignment.dx

```diff
@@ -0,0 +1,381 @@
+  To begin we need to be able to read in .wav files. To do this
+  we follow the wave file spec and define a header and the raw
+  data format.
+
+' [Wave Format Documentation](https://docs.fileformat.com/audio/wav/)
+  
+
+data WavHeader =
+     AsWavHeader {riff:(Fin 4=>Char) &
+                  size:Int &
+                  type:(Fin 4=>Char) &
+                  chunk:(Fin 4=>Char) &
+                  formatlength:Int &
+                  channels:Int & 
+                  samplerate:Int &
+                  bitspersample:Int &
+                  datasize:Int}
+                 
+data Wav = AsWav header:WavHeader dat:(List Int)
+
+' ### Binary Manipulation
+  We will need a couple additional binary conversion functions. (These are naive for now.)
+
+z = FToI (pow 2.0 15.0)
+def W32ToI (x : Word32): Int  =
+    y:Int = internalCast _ x
+    select (y <= z) y ((-1)*z + (y- z))
+       
+def W8ToW32 (x : Word8): Word32  = internalCast _  x
+
+def bytesToInt32 (chunk : Fin 4 => Byte) : Int =
+  [a, b, c, d] = map W8ToW32 chunk
+  W32ToI ((d .<<. 24) .|. (c .<<. 16) .|. (b .<<. 8) .|. a)
+
+
+def bytesToInt16 (chunk : Fin 2 => Byte) : Int =
+  [a, b] = map W8ToW32 chunk
+  W32ToI ((b .<<. 8) .|. a)
+
+
+' ### Parsing Types
+
+' We also add a couple additional parser helpers.
+
+def parseGroup (parser: Parser a) : (Parser (m => a)) = MkParser \h. for i. parse h parser
+parse4: Parser (Fin 4=>Char) = parseGroup parseAny
+parse2: Parser (Fin 2=>Char) = parseGroup parseAny
+
+def parse16 : Parser Int = MkParser \h.
+    bytesToInt16 $ parse h parse2
+
+def pString (x:String) : Parser Unit = MkParser \h.
+    (AsList n ls) = x
+    iter \i .
+         case i < n of
+              True ->
+                  _ = parse h (pChar ls.(i@_))
+                  Continue
+              False -> Done ()
+
+
+' ### Wave Parsing and IO
+
+' We define a parser to construct the header and to read in the data.
+
+wavparser : Parser Wav = MkParser \h.          
+              riff = parse h parse4
+              size = bytesToInt32 $ parse h parse4
+              type = parse h parse4
+              chunk = parse h parse4
+              _ = parse h parse16
+              formatlength = bytesToInt32 $ parse h parse4
+              channels = bytesToInt16 $ parse h parse2
+              samplerate = bytesToInt32 $ parse h parse4
+              _ = bytesToInt32 $ parse h parse4
+              _ = bytesToInt16 $ parse h parse2
+              bitspersample = bytesToInt16 $ parse h parse2
+              _ = parse h $ pString (AsList 4 ['d', 'a', 't', 'a'])
+              datasize = bytesToInt32 $ parse h parse4
+              x = parse h $ parseMany parse16
+              header = AsWavHeader {riff=riff, size=size,
+                     type=type, chunk=chunk,
+                     formatlength=formatlength,
+                     channels=channels, samplerate=samplerate,
+                     bitspersample=bitspersample,
+                     datasize=datasize}
```

Comment:
Oh nice. I knew I could do that for unpacking, didn't realize it worked for construction. 

It also wasn't clear to me the difference between a data of a record and a data of named fields. I went with record because the printed output was nicer.

Reply:
Regarding the record wildcard, that seems like it would maybe work in checking mode but not in inference mode? Not sure if there's a good way to restrict it to places where the set of locals you want is "obvious". I guess data-of-a-record is probably a special-enough case that we could do something like that there? But not sure about the general case.